src/App.js

```diff
@@ -1,39 +1,65 @@
 import './App.scss';
 import {
-  BrowserRouter as Router,
-  Switch,
-  Route,
-  Redirect
+    BrowserRouter as Router,
+    Switch,
+    Route,
+    Redirect
 } from "react-router-dom";
 import { useState, useEffect } from 'react';
+import { useDispatch } from "react-redux";
+import { UpdateProfile } from "./actions/actions"
 import firebase from "firebase";
 import LoginRoute from './components/LoginRoute'
 import MainRoute from './components/MainRoute'
 
 function App() {
 
-  const [user, setUser] = useState(null);
+    const [loaded, setLoaded] = useState(false);
+    const [fetchedUser, setFetchedUser] = useState(false);
+    const dispatch = useDispatch();
 
-  useEffect(() => {
-    firebase.auth().onAuthStateChanged(function (user) {
-      setUser(user);
-    });
-  }, []);
+    useEffect(() => {
+        var unsubscirbeProfileUpdates = null;
+        firebase.auth().onAuthStateChanged(function (user) {
+            setLoaded(true);
+            setFetchedUser(false);
 
-  return (
-    <div className="App">
-      <Router>
-        <Switch>
-          <Route exact path="/">
-            {!user ? <LoginRoute /> : <Redirect to="/tasks" />}
-          </Route>
-          <Route path="/:section">
-            {user ? <MainRoute /> : <Redirect to="/" />}
-          </Route>
-        </Switch>
-      </Router>
-    </div>
-  );
+            if (user) {
+                dispatch(UpdateProfile(null));
+                unsubscirbeProfileUpdates = firebase.firestore().collection("users").doc(user.uid)
+                    .onSnapshot(function (doc) {
```

Comment:
Looks like you are mixing UI & other (Network, Data) layers.
To follow `react-hooks` code style it is better to move such big hooks into the custom ones, e.g. it should looks like this:
```
// hooks/useAuthStateChanged.js
const useAuthStateChanged = () => {
      const [loaded, setLoaded] = useState(false);
      const [fetchedUser, setFetchedUser] = useState(false);
      const dispatch = useDispatch();

      useEffect(()=> {
          // logic
      }, []);

     return { loaded, fetchedUser };
}

export default useAuthStateChanged;

// src/App.js

import useAuthStateChanged from 'hooks/useAuthStateChanged'

function App() {
     const { loaded, fetchedUser } = useAuthStateChanged();

    ...
}
```